### PR TITLE
Allow BufferViews to have non-owning references to DeviceBuffers

### PR DESCRIPTION
**Pull request by gaaclarke**
issue: https://github.com/flutter/flutter/issues/157538

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or the PR is [test-exempt]. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by gaaclarke** [comment]
@jonahwilliams I started sketching out what it could look like to have raw pointers in DeviceBuffer.  I think I can make this work, especially if we can move DeviceBuffer to unique_ptr in a later PR.  The gist is that DeviceBuffer can have ownership or not depending if it's coming from the HostBuffer's DeviceBuffers or not.

**Comment by jonahwilliams** [comment]
That sounds like a good approach. off the top of my head, there is a DeviceBuffer::AsBufferView API that is problematic without the buffer view having ownership, but we can fix that by requiring an explicit track too.

File: impeller/renderer/backend/vulkan/render_pass_vk.cc [hunk 1]
```diff
@@ -399,27 +404,27 @@ bool RenderPassVK::SetIndexBuffer(BufferView index_buffer,
...
-        index_buffer_view.buffer;
-    if (!index_buffer) {
+    if (!index_buffer_view.GetBuffer()) {
       VALIDATION_LOG << "Failed to acquire device buffer"
                      << " for index buffer view";
       return false;
     }
 
-    if (!command_buffer_->Track(index_buffer)) {
+    std::shared_ptr<const DeviceBuffer> index_buffer =
```
Comment (jonahwilliams) [inline impeller/renderer/backend/vulkan/render_pass_vk.cc:418]: So we only track it if the buffer_view owns the device buffer reference? Oh and then we set the buffer view to use a raw_ptr as we know its tracked already...

Incredibly clever!

**Comment by gaaclarke** [comment]
The reason this doesn't work as-is right now is that if we stop tracking DeviceBuffers from the HostBuffer; when ~HostBuffer() is called it will schedule those DeviceBuffers in the HostBuffer to be deleted which will result in deleting the DeviceBuffer while the command buffer is referencing it.  We'd need to create some sort of mechanism where ~HostBuffer waits for all frames to render.  Or maybe in ~HostBuffer passing all the DeviceBuffers to the TrackedObjectsVK.

**Review by jonahwilliams — APPROVED** [review]
LGTM

**Comment by jonahwilliams** [comment]
reason for revert: merge issue.

**Pull request merged**